homeassistant/components/sensor/linky.py

```diff
@@ -5,88 +5,227 @@
-        client.close_session()
-        return
+    linky_data = LinkyData(username, password, timeout)
+    linky_data.update()
 
-    devices = [LinkySensor('Linky', client)]
-    add_entities(devices, True)
+    add_entities([LinkySensor(name, linky_data, peak_hours,
+                              peak_hours_cost, offpeak_hours_cost)])
+    return True
 
 
 class LinkySensor(Entity):
     """Representation of a sensor entity for Linky."""
 
-    def __init__(self, name, client):
+    def __init__(self, name, linky_data,
+                 peak_hours, peak_hours_cost,
+                 offpeak_hours_cost):
         """Initialize the sensor."""
         self._name = name
-        self._client = client
+        self._peak_hours = peak_hours
+        self._peak_hours_cost = peak_hours_cost
+        self._offpeak_hours_cost = offpeak_hours_cost
+        self._unit = KILOWATT_HOUR
+        self._icon = "mdi:flash"
+        self._lk = linky_data
         self._state = None
+        self._attributes = {}
+        self.update()
 
     @property
     def name(self):
         """Return the name of the sensor."""
         return self._name
 
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
     @property
     def state(self):
         """Return the state of the sensor."""
         return self._state
 
     @property
-    def unit_of_measurement(self):
-        """Return the unit of measurement."""
-        return 'kWh'
+    def icon(self):
+        """Return the icon of the sensor."""
+        return self._icon
 
     @Throttle(SCAN_INTERVAL)
     def update(self):
         """Fetch new state data for the sensor."""
+        _LOGGER.debug("Start of update of linky data")
+        self._lk.update()
+        if not self._lk.success:
+            self._state = STATE_UNAVAILABLE
+            self._attributes = []
+            return
+
+        self._state = self._lk.daily[1][CONSUMPTION]
+        self._attributes["halfhourly"] = [d[CONSUMPTION]
+                                          for d in self._lk.halfhourly]
+        self._attributes["daily"] = [d[CONSUMPTION]
+                                     for d in self._lk.daily]
+        self._attributes["peak_hours"] = sum([
+            d[CONSUMPTION]
+            if any([_between(h[0], h[1], d[TIME])
+                    for h in self._peak_hours])
+            else 0 for d in self._lk.halfhourly]) / 2
+        # From kW for 30 minutes to kWh
+        self._attributes["offpeak_hours"] = sum(
+            [0 if any([_between(h[0], h[1], d[TIME])
+                       for h in self._peak_hours])
+             else d[CONSUMPTION]
+             for d in self._lk.halfhourly]) / 2
+        # From kW for 30 minutes to kWh
+        self._attributes["peak_offpeak_percent"] = ((self._attributes
```

Comment:
Some users share on the forum to have divison by Zero here. Perhaps add some protections.